src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
     # pylint: disable=protected-access
     future.pimpl._clear()
     return status
+
+
+def _validate_keys(expected, given, keys_optional=False, allow_additional=False):
+    if not isinstance(expected, set):
+        expected = set(expected)
+    if not isinstance(given, set):
+        given = set(given)
+    if not keys_optional:
+        for req_key in expected.difference(given):
+            raise ValueError("Missing key ({})".format(req_key))
+    if not allow_additional:
+        for extraneous_key in given.difference(expected):
+            raise ValueError("Extraneous key ({})".format(extraneous_key))
+
+
+class Jobspec(object):
+    top_level_keys = set(["resources", "tasks", "version", "attributes"])
+
+    def __init__(self, resources, tasks, version, attributes=None):
+        if not isinstance(resources, collectionsAbc.Sequence):
+            raise TypeError("resources must be a sequence")
+        if not isinstance(tasks, collectionsAbc.Sequence):
+            raise TypeError("tasks must be a sequence")
+        if not isinstance(version, int):
+            raise TypeError("version must be an integer")
+        if attributes is not None and not isinstance(
+            attributes, collectionsAbc.Mapping
+        ):
+            raise TypeError("attributes must be a mapping")
+        elif version < 1:
+            raise ValueError("version must be >= 1")
+
+        self.jobspec = {
+            "resources": resources,
+            "tasks": tasks,
+            "attributes": attributes,
+            "version": version,
+        }
+
+        for res in self.resource_walk():
+            self._validate_resource(res)
+
+        for task in tasks:
+            self._validate_task(task)
+
+        if attributes is not None:
+            self._validate_attributes(attributes)
+
+    @classmethod
+    def from_yaml_stream(cls, yaml_stream):
+        jobspec = yaml.safe_load(yaml_stream)
+        _validate_keys(cls.top_level_keys, jobspec.keys())
+        return cls(**jobspec)
+
+    @classmethod
+    def from_file(cls, filename):
```

Comment:
I agree with @garlick, it would be good if this were from_yaml_file or similar for consistency.